design-docs/configuration-dsl.md

```diff
@@ -0,0 +1,376 @@
+linker between tool chain family (Clang/C2, NASM) to produce binaries for the current platform. When targeting more exotic platform, you can use
+preprocessor, compiler, and linker of toolchains for those platforms. It's called cross-compiling. The compiler runs on a host machine targeting an
+entirely different architecture and/or operating system. Cross-compilation is all about the linker and sometimes the compiler. It means you could
+mix and match preprocessor with other compiler and linker.
+
+Clang family also bring another twist as you can dump the intermediate language (IR), execute transformations on it and feed it back to the backend
+of the compiler to complete the compilation. MSVC is moving in this direction as well, and GCC also supports this workflow.
+
+The proposed solution here would be to allow users to access the tool chain resolver more directly as well as enabling them to declare custom tool
+chains. Tool chains can be located in the system PATH, at predefined locations, identified from the registry (Windows) or identified through COM
+object (Windows), fetched remotely, installed through unattended installers or package manager or manually configured.
+
+## Open Issues
+ - The source dialect & ABI is its own beast to support. It's suggested modeling it separately and use compiler version to match variant
+ - Should it be named `tool chain` (Gradle) or `toolchain` (Wikipedia)?
+ - Should the tool chain and the libraries contained within (prebuilt libraries) be two separate concepts? Meaning the toolchain resolver can be used by prebuilt dependencies but would be tied together.
+
+## DSL
+
+```
+toolChainResolvers {
+    visualStudio {
+        // additional configuration for visual studio resolver
+    }
+    myCustomResolver(MyCustomResolver) {
+        // configuration
+    }
+}
+
+toolChains {
+    myCustomGcc(Gcc) {
+        path '...'
+    }
+    myCustomVisualCpp(VisualCpp) {
+        installDir '...'
+    }
+    myCustomToolChain(ToolChain) {
+        tools.add(GccCppCompiler) {
+            it.dimensions architectures.x64, operatingSystems.windows
+            it.executable = file('/some/other/path')
+            it.withArguments { args ->
+                args << 'some-arg'
+            }
+        }
+
+        tools.add(GccLinker) {
+            it.dimensions architecture.x64, operatingSystem.windows
+            // ...
+        }
+
+        installedBy tasks.myCustomToolChainInstallTask
+
+        // TODO - Define how to add libraries
+        libraries.add(...) {
+            // added libraries can have a...
+            //   ...group, e.g. 'com.some.toolchain'
+            //   ...name, e.g. 'libc'
+            //   ...version (optional), e.g. '2.1'
+            // Adding a dependency without a version would depend on the latest or only version declared
+        }
+    }
+}
+
+repositories {
+    toolChains.myCustomToolChain
```

Comment:
This should be implicit. We should always use the available tool chains - and the selected tool - as candidate library providers.